jwt-authorizer/src/layer.rs

```diff
@@ -183,12 +184,68 @@ where
     }
 
     /// Build axum layer
+    #[deprecated(since = "0.10.0", note = "please use `to_layer()` instead")]
     pub async fn layer(self) -> Result<AsyncAuthorizationLayer<C>, InitError> {
         let val = self.validation.unwrap_or_default();
-        let auth = Arc::new(Authorizer::build(&self.key_source_type, self.claims_checker, self.refresh, val).await?);
-        Ok(AsyncAuthorizationLayer::new(auth, self.jwt_source))
+        let auth = Arc::new(
+            Authorizer::build(self.key_source_type, self.claims_checker, self.refresh, val, self.jwt_source).await?,
+        );
+        Ok(AsyncAuthorizationLayer::new(vec![auth]))
     }
 }
+
+#[async_trait]
+impl<C> IntoLayer<C> for JwtAuthorizer<C>
+where
+    C: Clone + DeserializeOwned + Send + Sync,
+{
+    async fn into_layer(self) -> Result<AsyncAuthorizationLayer<C>, InitError> {
+        let val = self.validation.unwrap_or_default();
+        let auth = Arc::new(
+            Authorizer::build(self.key_source_type, self.claims_checker, self.refresh, val, self.jwt_source).await?,
+        );
+        Ok(AsyncAuthorizationLayer::new(vec![auth]))
+    }
+}
+
+#[async_trait]
+impl<C, T> IntoLayer<C> for T
+where
+    T: IntoIterator<Item = JwtAuthorizer<C>> + Send + Sync,
+    C: Clone + DeserializeOwned + Send + Sync,
+{
+    async fn into_layer(self) -> Result<AsyncAuthorizationLayer<C>, InitError> {
+        let mut errs = Vec::<InitError>::new();
+        let mut auths = Vec::<Arc<Authorizer<C>>>::new();
+        let mut auths_futs: FuturesUnordered<_> = self
+            .into_iter()
+            .map(|a| {
+                Authorizer::build(
+                    a.key_source_type,
+                    a.claims_checker,
+                    a.refresh,
+                    a.validation.unwrap_or_default(),
+                    a.jwt_source,
+                )
+            })
+            .collect();
+
+        while let Some(a) = auths_futs.next().await {
+            match a {
+                Ok(res) => auths.push(Arc::new(res)),
+                Err(err) => errs.push(err),
+            }
+        }
+
+        if let Some(e) = errs.into_iter().next() {
+            // TODO: composite build error (containing all errors)
```

Comment:
This is fundamentally why i pushed building the authorizers back up into the user code ; So a crate user can handle the errors early and more easily. 

Even if this ends up returning some kind of composite error for all authorizers in the iterator handling that will be incredibly tedious, as the only way for a crate user to correlate it back to a specific authorizer would be to keep a parallel list of authorizing settings to tie it to.

Reply:
Ok, I think this is a good point, my last commit allows to build the authorizers separately (keeping the code backwards compatible). This could facilitate error management (errors per authorizer build).

 - jwt_authorizer::layer() -> backwards compatibility (async, returns Result)
 - built_authorizer(s).into_layer() -> no need to be async and can return a layer directly